nemo/lightning/pytorch/plugins/mixed_precision.py

```diff
@@ -84,14 +83,10 @@ def convert_optimizer(self, optimizer: Optimizer) -> Optimizer:
         """
         from nemo.core.optim import MainParamsOptimizerWrapper
 
-        if isinstance(optimizer, MainParamsOptimizerWrapper) or not self.amp_O2:
-            return optimizer
-
-        return MainParamsOptimizerWrapper(
-            optimizer,
-            fp32_grad_accum=True,
-            contiguous_grad_bucket=True,
-        )
+        optim_config = get_optim_config(optimizer)
```

Comment:
instead of asserts, could we change the optimizer args if self.precision is 16? this way we make the mixed precision plugin the only entry point
```suggestion
        if self.precision == "bf16-mixed":
          optimizer.mcore_optimizer.config.bf16 = True
        elif self.precision == "fp16-mixed":
          optimizer.mcore_optimizer.config.fp16 = True
```

Reply:
I'm leaning towards keeping the current code. I want to avoid modifying configs that were used to create objects (in this case the optimizer). This is mostly to catch misconfigurations.

Reply:
Yeah I'm not sure if those flags would be read before this function. @marcromeyn do you know? We could also do that in a followup PR